rfc/rfc-60/rfc-60.md

```diff
@@ -196,13 +195,75 @@ for metadata table to be populated.
 
 4. If there is an error reading from Metadata table, we will not fall back listing from file system.
 
-5. In case of metadata table getting corrupted or lost, we need to have a solution here to reconstruct metadata table
-from the files which distributed using federated storage. We will likely have to implement a file system listing
-logic, that can get all the partition to files mapping by listing all the prefixes under the `Table Storage Path`.
-Following the folder structure of adding table name/partitions under the prefix will help in getting the listing and
-identifying the table/partition they belong to.
+### Integration
+This section mainly describes how storage strategy is integrated with other components and how read/write
+would look like from Hudi side with object storage layout.
+
+We propose integrating the storage strategy at the filesystem level, specifically within `HoodieWrapperFileSystem`. 
+This way, only file read/write operations undergo path conversion and we can limit the usage of 
+storage strategy to only filesystem level so other upper-level components don't need to be aware of physical paths.
+
+This also mandates that `HoodieWrapperFileSystem` is the filesystem of choice for all upper-level Hudi components.
+Getting filesystem from `Path` or such won't be allowed anymore as using raw filesystem may not reach 
+to physical locations without storage strategy. Hudi components can simply call `HoodieMetaClient#getFs` 
+to get `HoodieWrapperFileSystem`, and this needs to be the only allowed way for any filesystem-related operation. 
+The only exception is when we need to interact with metadata that's still stored under the original table path, 
+and we should call `HoodieMetaClient#getRawFs` in this case so `HoodieMetaClient` can still be the single entry
+for getting filesystem.
+
+![](wrapper_fs.png)
+
+When conducting a read operation, Hudi would: 
```

Comment:
can we detail how any write operations will work? I guess it does the hashing and writes the object someplace else?

Reply:
This original idea is to register HoodieWrapperFileSystem on file writers like `HoodieSparkParquetWriter`. But with the recent addition of `HoodieStorage`, I believe there is an easier & cleaner way to achieve that. I'll leave this RFC as it is for now and will update this RFC with the idea of `HoodieStorage` in a separate PR